x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -40,34 +44,30 @@
  * A permission that is based on privileges for index related actions executed
  * on specific indices
  */
-public final class IndicesPermission implements Iterable<IndicesPermission.Group> {
+public final class IndicesPermission {
 
     public static final IndicesPermission NONE = new IndicesPermission();
+    private static final Automaton systemIndicesAutomaton = Automatons.patterns(SystemIndicesNames.indexNames());
 
-    private final Function<String, Predicate<String>> loadingFunction;
-
-    private final ConcurrentHashMap<String, Predicate<String>> allowedIndicesMatchersForAction = new ConcurrentHashMap<>();
+    private static final Logger logger = LogManager.getLogger();
 
+    private final ConcurrentMap<String, Predicate<String>> allowedIndicesMatchersForAction = new ConcurrentHashMap<>();
+    private final ConcurrentMap<Group, Automaton> indexGroupAutomatonCache = new ConcurrentHashMap<>();
```

Comment:
This is needed for the "has privilege" API. Otherwise we would avoid dealing with automatons in favor of dealing with predicates only (needed for the authorize use case). Since has privilege is called less frequently we built predicates lazily.

Reply:
I'd like to avoid another cache. We already have caching at the automata level. Maybe we could use this type of class within the group to implement the subset type functionality:

```java
    private static class NameMatcher implements Predicate<String> {

        private final Set<String> exactMatch;
        private final Set<String> nonExactMatch;
        private final Automaton automaton;
        private final CharacterRunAutomaton patternMatcher;

        private NameMatcher(Set<String> exactMatch, Set<String> nonExactMatch) {
            this.exactMatch = exactMatch;
            this.nonExactMatch = nonExactMatch;
            this.automaton = nonExactMatch.isEmpty() ? Automatons.EMPTY : Automatons.patterns(nonExactMatch);
            this.patternMatcher = Automatons.runAutomaton(automaton);
        }

        @Override
        public boolean test(String toTest) {
            return exactMatch.contains(toTest) || patternMatcher.run(toTest);
        }

        public boolean isSupersetOf(String value) {
            if (IndicesPermission.isIndexPattern(value)) {
                final Automaton valueAutomaton = Automatons.patterns(value);
                if (Operations.subsetOf(valueAutomaton, automaton)) {
                    return true;
                } else if (exactMatch.isEmpty() == false) {
                    // slow path
                    Automaton encompassingAutomaton =
                        Automatons.unionAndMinimize(Arrays.asList(automaton, Automatons.patterns(exactMatch)));
                    return Operations.subsetOf(valueAutomaton, encompassingAutomaton);
                }
                return false;
            } else {
                return test(value);
            }
        }

        @Override
        public String toString() {
            return "NameMatcher for " + Stream.concat(exactMatch.stream(), nonExactMatch.stream()).collect(Collectors.joining("|"));
        }
    }
```

Reply:
I am ok with not caching the complete automaton and only building the complete one in specific circumstances. I assume `has_privilege` is rare call. Alternatively, we could have a `LazyInitializable` member inside the `Group` (this is still caching, but not exposed). I would however put this all inside the `Group` and not create another class.

What do you say?

Reply:
I haven't thought about this as much as the 2 of you, so I might be missing something, but under what circumstances would an `exactMatch` index pattern, ever return true for a wildcard _has-privileges_ pattern?

`all` : `index-001` would never imply `read` : `index-*`

I think this proposed `NameMatcher` can be smart enough to implement this strategy on `isSupersetOf`
- if `value` in a plain string return `test(value)`
- else  return `Operations.subsetOf(Automatons.patterns(value), this.automaton)`

with whatever checks we need for `restricted_indices`